pinot-controller/src/main/java/com/linkedin/pinot/controller/api/upload/ZKOperator.java

```diff
@@ -0,0 +1,195 @@
+import com.linkedin.pinot.controller.api.resources.FileUploadPathProvider;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.filesystem.PinotFS;
+import com.linkedin.pinot.filesystem.PinotFSFactory;
+import java.io.File;
+import java.net.URI;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+import org.apache.helix.ZNRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The ZKOperator is a util class that is used during segment upload to set relevant metadata fields in zk. It will currently
+ * also perform the data move. In the future when we introduce versioning, we will decouple these two steps.
+ */
+public class ZKOperator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ZKOperator.class);
+  private PinotHelixResourceManager _pinotHelixResourceManager;
+  private ControllerConf _controllerConf;
+  private ControllerMetrics _controllerMetrics;
+
+  public ZKOperator(PinotHelixResourceManager pinotHelixResourceManager, ControllerConf controllerConf, ControllerMetrics controllerMetrics) {
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _controllerConf = controllerConf;
+    _controllerMetrics = controllerMetrics;
+  }
+
+  public void completeSegmentOperations(SegmentMetadata segmentMetadata, URI finalSegmentLocationURI, File currentSegmentLocation, boolean enableParallelPushProtection, HttpHeaders headers,
+      FileUploadPathProvider provider) throws Exception {
+    String rawTableName = segmentMetadata.getTableName();
+    String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(rawTableName);
+    String segmentName = segmentMetadata.getName();
+
+    ZNRecord znRecord = _pinotHelixResourceManager.getSegmentMetadataZnRecord(offlineTableName, segmentName);
+
+    // Brand new segment, not refresh, directly add the segment
+    if (znRecord == null) {
+      LOGGER.info("Adding new segment: {}", segmentName);
+      try {
+        moveSegmentToPermanentDirectory(currentSegmentLocation, finalSegmentLocationURI);
+        LOGGER.info("Moved segment {} from temp location {} to {}", segmentName, currentSegmentLocation.getAbsolutePath(), finalSegmentLocationURI.getPath());
+      } catch (Exception e) {
+        LOGGER.error("Could not move segment {} from table {} to permanent directory", segmentName, rawTableName);
+        throw new RuntimeException(e);
+      }
+      if (finalSegmentLocationURI.getScheme().equals("file")) {
+        String downloadUrl = ControllerConf.constructDownloadUrl(rawTableName, segmentName, provider.getVip());
+        _pinotHelixResourceManager.addNewSegment(segmentMetadata, downloadUrl);
+
+      } else {
+        _pinotHelixResourceManager.addNewSegment(segmentMetadata, finalSegmentLocationURI.toString());
+      }
+      return;
+    }
+
+    LOGGER.info("Segment {} already exists, refreshing if necessary", segmentName);
```

Comment:
Will be really nice if we can break parts of the code below into different methods to make this method shorter. Or, sub-class of ZkOperator class.